### ISSUE TITLE
@wordpress/env: Loopback requests fail

### ISSUE THREAD

**Issue by noisysocks** [WordPress/gutenberg#20569]
**Describe the bug**
WordPress isn't able to complete a loopback request when running under `wp-env`.

**To reproduce**
1. `npx wp-env start`
2. Navigate to http://localhost:8888/wp-admin/site-health.php

There's two errors related to this:

- The REST API encountered an error
- Your site could not complete a loopback request

**Expected behavior**
There should only be one critical error (the one about debug mode.)

**Additional context**
WordPress is trying to make a request to `http://localhost:8888` from the underlying Docker container but this fails because Apache is listening on port 80. One solution would be to set up the `docker-compose.yml` config to map `localhost` to the host machine.

**Comment by tbclark3** [comment]
I am dealing with the same issue.  Wordpress attempts to make a loopback request to its own hostname (not to localhost) on port 443 which fails because the container doesn't provide SSL.  If there were a way to force loopback requests to use http instead of https, that might solve the problem, but I haven't found one yet.

**Comment by adrianduffell** [comment]
This issue seems to impact WP Cron for me. It looks like the scheduled events can't run because it uses loopback requests to create a background process.

I am able to work around this with the [`ALTERNATE_WP_CRON` config](http://wordpress-hackers.1065353.n5.nabble.com/ALTERNATE-WP-CRON-Is-it-worth-it-tp39843p39846.html).

```js
// .wp.env.json

"config": {
  "ALTERNATE_WP_CRON": true
}
```

**Comment by glendaviesnz** [comment]
Two possible solutions to this that I have found are:

- Access wp-env instance through an external domain tunnel like https://ngrok.com/ in which case the callback will be made to that domain rather than localhost.
- Add a hosts entry on the local machine for `127.0.0.1 host.docker.internal` and access wp-env on `http://host.docker.internal:8888` - within a docker container host.docker.internal resolves to itself.

For both of the above solutions you will need to change your WP config/settings to accept that domain name instead of the default localhost that wp-env sets.

**Comment by miminari** [comment]
I reproduced this at @wordpress/env@4.0.4 in M1 Mac

**Comment by WraithKenny** [comment]
Somewhat related, I'm using an environment that has a custom domain (wordpress.org.test) and a locally issued SSL cert, and the problem exists. The root cause is that the loop-back request is running the http request with `sslverify` turned on, which will of course fail if there's no ssl, or if it's locally issued.

To fix this, I mount a must-use plugin (to disable `sslverify`) with the following contents:
```php
<?php
// Fix loopback on local development.
add_action( 'init', function() {
	if ( ! defined( 'WP_DEBUG' ) || ! WP_DEBUG ) {
		return;
	}

	// Only run if the site is on a local TLD.
	$host = wp_parse_url( site_url() )['host'];
	$parts = explode( '.', $host );
	$tld = end( $parts );
	if ( ! in_array( $tld, [ 'test', 'example', 'invalid', 'localhost' ], true ) ) {
		return;
	}

	add_filter( 'http_request_args', function( $parsed_args, $url ) {

		$request_host = wp_parse_url( $url )['host'];
		$host         = wp_parse_url( site_url() )['host'];

		// Loopback request.
		if ( $host === $request_host ) {
			$parsed_args['sslverify'] = false;
		}

		return $parsed_args;
	}, 10, 2 );
});
```
`./dev/mu-plugins/fix-loopback.php:/var/www/html/wp-content/mu-plugins/fix-loopback.php`

**Comment by WraithKenny** [comment]
The question is, should this mu-plugin be offered and supported by default in `wp-env` to fix this bug.

**Comment by jmichaelward** [comment]
> This issue seems to impact WP Cron for me. It looks like the scheduled events can't run because it uses loopback requests to create a background process.
> 
> I am able to work around this with the [`ALTERNATE_WP_CRON` config](http://wordpress-hackers.1065353.n5.nabble.com/ALTERNATE-WP-CRON-Is-it-worth-it-tp39843p39846.html).
> 
> ```js
> // .wp.env.json
> 
> "config": {
>   "ALTERNATE_WP_CRON": true
> }
> ```

@adrianduffell  I'm wondering whether you're able to elaborate on your workaround for this a bit more? I've encountered this same issue while working on a plugin I'm developing that registers a custom cron task, and I've attempted to test the cron using the following commands:

- `wp-env run cli bash` - to shell into the container
- `wp cron test <my_event_name>` - to test the event.

This results in this error, whether or not `ALTERNATE_WP_CRON` is set to true in `.wp-env.json`:

```
Warning: The ALTERNATE_WP_CRON constant is set to true. WP-Cron spawning is not asynchronous.
Error: WP-Cron spawn failed with error: cURL error 7: Failed to connect to localhost port 8888 after 0 ms: Connection refused
```

It would be wonderful if we're able to find a solution for this issue. I've just started using wp-env for the first time and it's a terrific utility.

**Comment by ryelle** [comment]
This issue also prevents loading editor styles from plugins (or themes if you're not paying attention).

For example, I have a plugin that sets up some font files, and it's added to the editor by `add_editor_style( plugins_url( 'style.css', __FILE__ ) );`

When the editor styles are loaded, it tries to `wp_remote_get` the file contents, but it's prevented by "cURL error 7: Failed to connect to localhost port 8888: Connection refused". 

I had also added the theme's editor styles with `add_editor_style( get_template_directory_uri() . '/build/editor.css' );` causing the same issue, but that's easily fixed by switching to `add_editor_style( '/build/editor.css' );`

**Comment by Sam-Smyth** [comment]
@noisysocks It looks like this was never resolved and I'm running into this issue. Your solution of editing the `docker-compose.yml` file doesn't seem to work since `wp-env start` overwrites that file, right? That's what I'm finding.

**Comment by noisysocks** [comment]
Yeah, I was referring to making it so that `wp-env` writes a `docker-compose.yml` file that maps `localhost` to the host machine.

I don't use `wp-env` anymore (lately I've been trying https://ddev.com) so haven't been prioritising work on it. Feel free to open a PR in this repo containing a fix.

**Comment by mikey242** [comment]
I was having this issue too and one solution that I settled on was to change the default port number from `8888` to `80`. This obviously won't work if you are using port `80` for another service but it is a relatively simple fix to implement.

You can either add `"port": 80` to your `.wp-env.json` file or prepend your `wp-env` command like this: `WP_ENV_PORT=80 wp-env start`

**Comment by celtislab** [comment]
I created a plugin to prevent loopback request errors in WordPress environments that use Docker, such as wp-env.

[localhost2host-docker-internal](https://wordpress.org/plugins/localhost2host-docker-internal/)

Give it a try!

**Comment by rmorse** [comment]
Thanks @celtislab , I dropped this into my `mu-plugins` and its working great right now.

**Comment by BrianHenryIE** [comment]
Potentially, wp-env could script some changes to the Apache config as described here: https://github.com/docker-library/wordpress/issues/611#issuecomment-1378316911

I tried to manually do it but it didn't seem to work

```
sed -i 's/:80/:80 *:8888/g' /etc/apache2/sites-enabled/000-default.conf
sed -i 's/80/80\nListen 8888/g' /etc/apache2/ports.conf
```

**Comment by collinr3** [comment]
I got it to work using _ as the delimiter instead of / ie
`sed -i 's_80_80\nListen 8888_g' /etc/apache2/ports.conf`
